### ISSUE TITLE
GRPO unbalanced-memory

### ISSUE THREAD

**Issue by mdy666** [huggingface/trl#2805]
### Reproduction

the memory in each rank(0-6) is not same, and i find when the train steps increase, the memory will increase much
step 0  use the origin code

![Image](https://github.com/user-attachments/assets/cea459ef-a998-4707-b78e-0d753ba32481)

Then i write a efficient grpo loss kernel by triton。
step 0

![Image](https://github.com/user-attachments/assets/de360884-491a-4c47-917e-78a7fb579207)

step 5

![Image](https://github.com/user-attachments/assets/2d148490-bb44-417c-8d7d-7befa2d853e7)

step 20

![Image](https://github.com/user-attachments/assets/1bd7930a-ce78-4183-9d6b-1e6e9ebcc6fd)

### System Info

trl = 0.14.0
torch = 2.5.1+cuda12.4
vllm = 0.7.1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by mdy666** [comment]
I do lots of exps, I finally find it's caused by accuracy_reward
if set 
```python
def accuracy_reward(completions, solution, **kwargs):
    return [0.5 for _ in range(len(completions))]
```
,with the steps increase, the memory in my code (use triton write the grpo loss), the memory does not increase. Why I don't know why this would happen?

**Comment by nomadlx** [comment]
Is it convenient for you to provide your implementation? Currently, on an 8*80G graphics card, I can only run a 7B-sized model, and the maximum length for completion can only be set to 1024. This length is far from sufficient for an inference model. I don't quite understand why it leads to such high VRAM consumption.

**Comment by mdy666** [comment]
> Is it convenient for you to provide your implementation? Currently, on an 8*80G graphics card, I can only run a 7B-sized model, and the maximum length for completion can only be set to 1024. This length is far from sufficient for an inference model. I don't quite understand why it leads to such high VRAM consumption.

Sure! Here is my code. please click a star, thank you.
https://github.com/mdy666/mdy_triton/tree/main/others/grpo

**Comment by leonardtang** [comment]
> I do lots of exps, I finally find it's caused by accuracy_reward if set
> 
> def accuracy_reward(completions, solution, **kwargs):
>     return [0.5 for _ in range(len(completions))]
> with the steps increase, the memory in my code (use triton write the grpo loss) does not increase. But I don't know why this would happen?

You're saying that `accuracy_reward` is eating up GPU memory?

**Comment by mdy666** [comment]
> > I do lots of exps, I finally find it's caused by accuracy_reward if set
> > def accuracy_reward(completions, solution, **kwargs):
> > return [0.5 for _ in range(len(completions))]
> > with the steps increase, the memory in my code (use triton write the grpo loss) does not increase. But I don't know why this would happen?
> 
> You're saying that `accuracy_reward` is eating up GPU memory?

this sound incredible，but my exps tell me it caused by the math_verfiy package. The GPU memory will increase from 60G to 120G

**Comment by albertvillanova** [comment]
Hey @mdy666, sorry for the very slow reply here — we're catching up on the backlog. 

The trouble is there's nothing TRL-specific to act on as filed: the "Reproduction" is screenshots plus your own external Triton GRPO-loss kernel rather than stock `GRPOTrainer`, and "System Info" is three version strings rather than `trl env` output, so we can't isolate the growth to TRL. On the later finding — `math_verify` runs on CPU over strings, so it can't directly allocate the per-step GPU logits memory you describe (≈ `2*bs*num_generations*seq_len*vocab_size`); that growth almost certainly lives in the custom kernel holding onto logits. For reference, on current TRL the reward moved to `trl/rewards/accuracy_rewards.py` and now guards `math_verify` with timeouts + a main-thread check (https://github.com/huggingface/trl/blob/7904dc11cf8049c70e622e9ab3200ee5d9aa044d/trl/rewards/accuracy_rewards.py#L71-L78), and any residual sympy memory behavior is upstream in math-verify. 

Closing for now — very happy to re-open if you can share a minimal repro on stock `GRPOTrainer` (no custom kernel) with a current TRL version and `trl env` output.
